org.osgi.test.junit5/src/main/java/org/osgi/test/junit5/config/ConfigurationExtension.java

```diff
@@ -0,0 +1,156 @@
+							resolveReturnValue(configAnnotation, bundleContext, extensionContext);
+						});
+				}
+				Configuration configAnnotation = element.getAnnotation(Configuration.class);
+				if (configAnnotation != null) {
+					resolveReturnValue(configAnnotation, bundleContext, extensionContext);
+				}
+			});
+	}
+
+	private static org.osgi.service.cm.Configuration resolveReturnValue(Configuration configurationAnnotation,
+		BundleContext bundleContext, ExtensionContext extensionContext) {
+		ServiceReference<ConfigurationAdmin> sref = null;
+		try {
+			sref = bundleContext.getServiceReference(ConfigurationAdmin.class);
+			if (sref == null) {
+				throw new IllegalStateException("The configuration admin service cannot be found.");
+			}
+
+			ConfigurationAdmin configurationAdmin = bundleContext.getService(sref);
+			if (configurationAdmin == null) {
+				throw new IllegalStateException("The configuration admin service cannot be found.");
+			}
+
+			org.osgi.service.cm.Configuration configuration = null;
+
+			if (configurationAnnotation.factory()) {
+
+				configuration = configurationAdmin.getFactoryConfiguration(configurationAnnotation.pid(), null);
+			} else {
+				configuration = configurationAdmin.getConfiguration(configurationAnnotation.pid(), null);
+			}
+
+			if (configurationAnnotation.properties() != null) {
+				Map<String, ?> map = Stream.of(configurationAnnotation.properties())
+					.collect(Collectors.toMap((e) -> e.key(), (e) -> e.value()));
+				configuration.update(Dictionaries.asDictionary(map));
+			}
+			return configuration;
+
+		} catch (Exception e) {
+			throw new ExtensionConfigurationException(e.getMessage(), e);
+		} finally {
+			if (bundleContext != null && sref != null) {
+				bundleContext.ungetService(sref);
+			}
+		}
+	}
+
+	static void assertValidFieldCandidate(Field field) {
+		if (Modifier.isFinal(field.getModifiers()) || Modifier.isPrivate(field.getModifiers())
+			|| Modifier.isStatic(field.getModifiers())) {
+			throw new ExtensionConfigurationException("@" + Configuration.class.getSimpleName() + " field ["
+				+ field.getName() + "] must not be final, private or static.");
+		}
+	}
+
+	@Override
+	public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+		throws ParameterResolutionException {
+
+		BundleContext bundleContext = FrameworkUtil.getBundle(extensionContext.getRequiredTestClass())
+			.getBundleContext();
+
+		Optional<Configuration> configAnnotation = parameterContext.findAnnotation(Configuration.class);
+		Parameter parameter = parameterContext.getParameter();
+
+		return resolveReturnValue(configAnnotation.get(), bundleContext, extensionContext);
+	}
+
+	@Override
+	public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+		throws ParameterResolutionException {
+
+		if (parameterContext.isAnnotated(Configuration.class)) {
+			return true;
+		}
+		return false;
+	}
```

Comment:
Best practice (I'm using AssertJ's `SoftAssertionsExtension` for reference here as it was written by one of the JUnit 5 team) is for `supportsParameter` to also do the sanity checking if you know it's supposed to be one of your parameters, and throw a `ParameterResolutionException` if it doesn't pass validation. Again, have a look at `BundleContextExtension` for an example of how to do this (and how to write unit tests for it too).